src/client.rs

```diff
@@ -1,107 +1,132 @@
 
 use crate::config;
 use crate::remote;
 
+type TunSink = SplitSink<Framed<AsyncDevice, TunPacketCodec>, TunPacket>;
+
 pub struct Client {
-    tunnels: Vec<Arc<Tunnel>>,
-    receiver: mpsc::Receiver<TunPacket>,
-    remote: Arc<remote::Remote>,
+    tunnels: Arc<Vec<Tunnel>>,
+    remote_receiver: mpsc::Receiver<(String, Bytes)>,
+    local_addr: Ipv4Addr,
+    sink: TunSink,
 }
 
-#[derive(Clone)]
 struct Tunnel {
     id: Vec<u8>,
     topic: String,
+    topic_base: String,
     bind_addr: std::net::Ipv4Addr,
 }
 
 impl Client {
     pub async fn new(config: &config::Config) -> Self {
+        log::info!(
+            "Creating tun at {:?} network {:?}",
+            config.driver.local_addr,
+            config.driver.tun.netmask
+        );
+        let mut tun_config = tun::Configuration::default();
+        tun_config.address(config.driver.local_addr);
+        tun_config.destination(config.driver.local_addr);
+        tun_config.netmask(config.driver.tun.netmask);
+
+        #[cfg(target_os = "linux")]
+        tun_config.platform(|tun_config| {
+            tun_config.packet_information(true);
+        });
+
+        tun_config.up();
+
+        let dev = tun::create_as_async(&tun_config).expect("Tunnel");
+        let (sink, mut stream) = dev.into_framed().split();
+
+        let ip_network =
+            Ipv4Network::with_netmask(config.driver.local_addr, config.driver.tun.netmask)
+                .expect("A proper ip network");
         let mqtt_options = config.broker_mqtt_options();
 
-        let remote = remote::Remote::new(&mqtt_options, Vec::new());
-        let arc_remote = Arc::new(remote);
+        let (remote, remote_receiver) = remote::Remote::new(&mqtt_options, Vec::new());
 
         let client_config = config
             .client
             .as_ref()
             .expect("Client config to be non-null");
 
-        let (sender, receiver) = mpsc::channel(128);
         let mut tunnels = Vec::with_capacity(client_config.tunnels.len());
         let mut rng = thread_rng();
 
         for client_tunnel_config in &client_config.tunnels {
-            let mut tun_config = tun::Configuration::default();
-            tun_config.address(client_tunnel_config.bind_addr.clone());
-            if let Some(driver_tun_config) = &config.driver.tun {
-                tun_config.netmask(driver_tun_config.netmask.clone());
-            }
-            tun_config.up();
-
-            #[cfg(target_os = "linux")]
-            tun_config.platform(|tun_config| {
-                tun_config.packet_information(true);
-            });
-
-            let dev = tun::create_as_async(&tun_config).unwrap();
-            let mut stream = dev.into_framed();
-
             let random_id: Vec<u8> = (&mut rng)
-                .sample_iter(Standard)
+                .sample_iter(Alphanumeric)
                 .take(client_tunnel_config.id_length)
                 .collect();
-
             let base64_id = general_purpose::STANDARD.encode(&random_id);
```

Comment:
I haven't read the docs on `STANDARD.encode` but the previous encoding scheme uses the URL-safe variant of Base64 (rfc3548 section 4) with no padding at the end (rfc3548 section 2.2). eg: `0xFFFFAAAA` becomes `__-qqg` instead of `//+qqg==`

Reply:
Ok let me use the [URL_SAFE version](https://docs.rs/base64/latest/base64/engine/general_purpose/constant.URL_SAFE.html)

Reply:
Seems there;s an option called [NO_PAD](https://docs.rs/base64/latest/base64/engine/general_purpose/constant.NO_PAD.html) too

Reply:
I think padding is fine, it's the "+" charachter that's illegal as a MQTT topic that I wanted to avoid. Lets use [URL_SAFE_NO_PAD](https://docs.rs/base64/latest/base64/engine/general_purpose/constant.URL_SAFE_NO_PAD.html#)